### PR TITLE
COIOS-755: Credit card field cursor fix

### PR DESCRIPTION
**Pull request by goergisn**
# Summary
- Allowing editing of the credit card number while maintaining the cursor position

# Demo
https://github.com/user-attachments/assets/b4994ac6-a2fd-4e64-8ca7-43274917a615

# Release notes

<fixed> 
In the **Card number** field, the shopper can now move the cursor and edit the number where the cursor is positioned.
</fixed>

# Ticket

<ticket>
COIOS-755
</ticket>

### CODE REVIEW

File: AdyenCard/Form/FormCardNumberItem.swift [hunk 1]
```diff
@@ -105,6 +105,51 @@ internal final class FormCardNumberItem: FormTextItem, AdyenObserver {
...
+    @discardableResult
+    internal func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        
+        guard
+            let text = textField.text,
+            let textRange = Range(range, in: text),
+            let selectedTextRange = textField.selectedTextRange
+        else { return true }
+        
+        let editingDirection = string.count - range.length
```
Comment (atmamont) [inline AdyenCard/Form/FormCardNumberItem.swift]: It's a bit confusing to see a direction being `Int`

Comment (goergisn) [reply AdyenCard/Form/FormCardNumberItem.swift]: Agree - renamed 👌

File: AdyenCard/Form/FormCardNumberItem.swift [hunk 2]
```diff
@@ -105,6 +105,51 @@ internal final class FormCardNumberItem: FormTextItem, AdyenObserver {
...
+        let formattedText = formatter?.formattedValue(for: sanitizedText) ?? sanitizedText
+        
+        let oldCursorOffset = textField.offset(from: textField.beginningOfDocument, to: selectedTextRange.end)
+        
+        let isAdding = formattedText.count > text.count
+        
+        let oldNumberOfSpacesBeforeCursor = text.numberOfSpaces(beforeOffset: oldCursorOffset)
+        
+        let newNumberOfSpacesBeforeCursor = formattedText.numberOfSpaces(
+            beforeOffset: oldCursorOffset + replacementLength + (isAdding ? 1 : 0)
```
Comment (nauaros) [inline AdyenCard/Form/FormCardNumberItem.swift]: I found this line a bit tricky to understand. Maybe creating a variable for `beforeOffset` with a meaningful name can help. Just a suggestion.

**Review by nauaros — DISMISSED** [review]

**Review by nauaros — APPROVED** [review]

File: AdyenCard/Form/FormCardNumberItem.swift [hunk 3]
```diff
@@ -165,10 +211,44 @@ internal final class FormCardNumberItem: FormTextItem, AdyenObserver {
...
+        range: NSRange,
+        replacementString: String,
+        in text: String,
+        isDeletingSingleCharacter: Bool
+    ) -> Int {
+        // Special case to allow "deleting" a space
+        // (can only be triggered when the user manually moves the cursor)
+        //
+        // 1234 5678 |310 // Deleting a character
+        if range.length == 1, replacementString.isEmpty, isDeletingSingleCharacter {
```
Comment (erenbesel) [inline AdyenCard/Form/FormCardNumberItem.swift:227]: Hmm nice! What about deleting multiple chars at once by dragging/selecting?

Comment (goergisn) [reply AdyenCard/Form/FormCardNumberItem.swift:227]: That already works - this was just a fix for a very edge-case where the user moves the cursor in front of a comma.

Reason behind it is that there are no "spaces" in a card number number - it's "just visual" formatting

**Review by erenbesel — APPROVED** [review]

**Pull request merged**